File: python/986. [M] Interval List Intersections.py
# https://leetcode-cn.com/problems/interval-list-intersections/

# You are given two lists of closed intervals, firstList and
# secondList, where firstList[i] = [starti, endi] and
# secondList[j] = [startj, endj]. Each list of intervals
# is pairwise disjoint and in sorted order.
#
# Return the intersection of these two interval lists.
#
# A closed interval [a, b] (with a < b) denotes the set of
# real numbers x with a <= x <= b.
#
# The intersection of two closed intervals is a set of real
# numbers that are either empty or represented as a closed
# interval. For example, the intersection of [1, 3] and [2, 4]
# is [2, 3].
#
# 0 <= firstList.length, secondList.length
#
# 来源：力扣（LeetCode）
# 链接：https://leetcode-cn.com/problems/interval-list-intersections
# 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。

class Solution:
    def intervalIntersection(self, firstList: list[list[int]],
                             secondList: list[list[int]]) -> list[list[int]]:

        def findOverlap(l: list[list[int]]) -> list[int]:
            l.sort(key=lambda x: x[0])

            # -----
            #       ---
            if l[0][1] < l[1][0]: return []

            # ------
            #  ---
            if l[0][1] >= l[1][1]: return l[1]

            # ----
            #   ----
            return [l[1][0], l[0][1]]

        res = []
        ptr1, ptr2 = 0, 0
        while ptr1 < len(firstList) and ptr2 < len(secondList):
            itv1, itv2 = firstList[ptr1], secondList[ptr2]
            overlap = findOverlap([itv1, itv2])
            if len(overlap) != 0: res.append(overlap)

            if itv1[1] < itv2[1]: ptr1 += 1
            else: ptr2 += 1

        return res



if __name__ == '__main__':
    sol = Solution()
    # print(sol.merge([[1,2]]))
    print(sol.intervalIntersection([[1,2],[3,5],[6,7]], [[2,3]]))
    # assert(sol.rob([2,4]) == 4)
    # assert(sol.rob([2,1,4,8]) == 9)
    # assert(sol.rob([2,1,4,8,9]) == 13)

File: cpp/986. [M] Interval List Intersections.cpp
#include <vector>  
#include <iostream>
#include <math.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
        int size1 = firstList.size(), size2 = secondList.size();
        vector<vector<int>> intersections = {};
        int i1 = 0, i2 = 0;
        while (i1 < size1 && i2 < size2) {
            vector<int> itv1 = firstList[i1], itv2 = secondList[i2];
            // if no overlap
            if (itv1[1] < itv2[0]) {
                i1++;
                continue;
            }
            if (itv2[1] < itv1[0]) {
                i2++;
                continue;
            }
            // if there is overlap
            intersections.push_back({max(itv1[0], itv2[0]), min(itv1[1], itv2[1])});
            if (itv1[1] < itv2[1])
                i1++;
            else
                i2++;
        }
        return intersections;
    }
};

int main() {
    Solution sol = Solution();
}